ibm/resource_ibm_securitygroup_rule.go

```diff
@@ -0,0 +1,253 @@
+				Type:         schema.TypeString,
+				Optional:     true,
+				Description:  "IP version IPv4 or IPv6",
+				Default:      "IPv4",
+				ForceNew:     true,
+				ValidateFunc: validateSecurityRuleEtherType,
+			},
+			"port_range_min": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				ForceNew: true,
+			},
+			"port_range_max": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				ForceNew: true,
+			},
+			"protocol": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Description:  "icmp, tcp or udp",
+				ForceNew:     true,
+				ValidateFunc: validateSecurityRuleProtocol,
+			},
+			"security_group_id": {
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func findMatchingRule(securitygroupid int, rule *datatypes.Network_SecurityGroup_Rule,
+	service services.Network_SecurityGroup) (*datatypes.Network_SecurityGroup_Rule, error) {
+
+	var filters []filter.Filter
+	if rule.PortRangeMax != nil {
+		filters = append(filters, filter.Path("rules.portRangeMax").Eq(rule.PortRangeMax))
+	}
+	if rule.PortRangeMin != nil {
+		filters = append(filters, filter.Path("rules.portRangeMin").Eq(rule.PortRangeMin))
+	}
+
+	filters = append(filters, filter.Path("rules.direction").Eq(rule.Direction))
+
+	if rule.Ethertype != nil {
+		filters = append(filters, filter.Path("rules.ethertype").Eq(rule.Ethertype))
+	}
+	if rule.Protocol != nil {
+		filters = append(filters, filter.Path("rules.protocol").Eq(rule.Protocol))
+	}
+
+	rules, err := service.Filter(filter.Build(filters...)).Id(securitygroupid).GetRules()
+	if err != nil {
+		return nil, fmt.Errorf("Error fetching information for Security Group Rule: %s", err)
+	}
+
+	if len(rules) == 0 {
+		return nil, nil
+	}
+	return &rules[0], nil
+}
+
+func resourceIBMSecurityGroupRuleCreate(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	service := services.GetNetworkSecurityGroupService(sess)
+
+	securitygroupid := d.Get("security_group_id").(int)
+
+	sgrule := datatypes.Network_SecurityGroup_Rule{}
+
+	sgrule.Direction = sl.String(d.Get("direction").(string))
+
+	if d.Get("ether_type").(string) != "" {
+		sgrule.Ethertype = sl.String(d.Get("ether_type").(string))
+	}
+
+	if _, ok := d.GetOk("port_range_min"); ok {
+		sgrule.PortRangeMin = sl.Int(d.Get("port_range_min").(int))
+
+	}
+
+	if _, ok := d.GetOk("port_range_max"); ok {
+		sgrule.PortRangeMax = sl.Int(d.Get("port_range_max").(int))
+
+	}
+
+	if d.Get("protocol").(string) != "" {
+		sgrule.Protocol = sl.String(d.Get("protocol").(string))
+	}
+
+	// if only one of min/max is provided, set the other one to the provided
+	if sgrule.PortRangeMin != nil && sgrule.PortRangeMax == nil {
+		sgrule.PortRangeMax = sgrule.PortRangeMin
+	}
+	if sgrule.PortRangeMax != nil && sgrule.PortRangeMin == nil {
+		sgrule.PortRangeMin = sgrule.PortRangeMax
+	}
+
+	if sgrule.PortRangeMin != nil {
+		if *sgrule.PortRangeMax < *sgrule.PortRangeMin {
```

Comment:
FYI, portRangeMin and portRangeMax when used with the protocol ICMP are reused as the ICMP type and code respectively. So allowing ICMP ping (type 8, code 0) would be protocol ICMP, portRangeMin 8, portRangeMax 0

We do have appropriate error checking in the API for this, so do you want to lean on that error checking instead of reimplementing it here?